src/ring_discord_auth/core.clj

```diff
@@ -81,54 +43,6 @@
                  (.get ^ByteBuffer input result)
                  result)))
 
-(defn encode
-  "Encodes the given string to a byte array using the given charset/encoding.
-
-  Returns `nil` if the charset is not available or if it doesn't support encoding."
-  {:test (examples encode Arrays/equals
-           ["Hello, world!", "utf8" #_=> (byte-array [0x48 0x65 0x6c 0x6c 0x6f 0x2c 0x20 0x77 0x6f 0x72 0x6c 0x64 0x21])]
-           ["Another one", "unknown?" #_=> nil])}
-  ^bytes [^String str ^String charset-name]
-  (if-let-all [^Charset cs (try (Charset/forName charset-name) (catch UnsupportedCharsetException _ nil) (catch IllegalCharsetNameException _ nil))
-               ^CharsetEncoder encoder (try (.newEncoder cs) (catch UnsupportedOperationException _ nil))]
-    (when (.canEncode encoder str)
-      (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))
-
-(defn authentic?
-  "Checks whether a signature is authentic, given a message and a public key.
-
-  This function has 2 arities: a general purpose one that simply delegates to a crypto library and one tailored to what Discord provides.
-
-  The general purpose arity takes a signature, a message and a public key (all of which must be byte arrays) and checks the authenticity.
-
-  The other arity takes a signature (byte array or hex string), public key (byte array or hex string), body (byte array or string), timestamp (byte array or string).
-  It combines timestamp and body and uses that as the message.
-
-  Returns `true` if the message is authentic, `false` if not."
-  ([signature-bytes message-bytes public-key-bytes]
-   (try
-     (sign/verify signature-bytes message-bytes public-key-bytes)
-     true
-     (catch RuntimeException _ false)))
-  ([signature body timestamp public-key charset-name]
-   (if-let-all [signature-bytes (cond-> signature (not (bytes? signature)) hex->bytes)
-                public-key-bytes (cond-> public-key (not (bytes? public-key)) hex->bytes)
-                ^bytes body-bytes (cond-> body (string? body) (encode body charset-name))
-                ^bytes timestamp-bytes (cond-> timestamp (string? timestamp) (encode timestamp charset-name))
-                message-bytes (byte-array (+ (alength timestamp-bytes) (alength body-bytes)))]
-     (do
-       (System/arraycopy timestamp-bytes 0 message-bytes 0 (alength timestamp-bytes))
-       (System/arraycopy body-bytes 0 message-bytes (alength timestamp-bytes) (alength body-bytes))
-       (authentic? signature-bytes message-bytes public-key-bytes))
-     false)))
```

Comment:
My problem with the current signature is the `System/arraycopy` you need to make to combine the two byte arrays, if you had then as string before they are bytes you can just merge with an str function and then convert to a byte array witch makes a very the code easier to read.

Reply:
Since the verification uses byte arrays in the end, that is what `authentic?` coerces to. This has 3 benefits:
- It's fast
- It saves unnecessary (and possibly error-prone) conversions
- It allows for different combinations of input types without additional overhead

It may not be the nicest to read, but performance was one thing I wanted to focus on right away.

Reply:
:+1: